test_runner/src/main/kotlin/ftl/run/AndroidTestRunner.kt

```diff
@@ -17,66 +18,69 @@ import kotlinx.coroutines.coroutineScope
-        repeat(runCount) {
-            repeat(shardCount) { testShardsIndex ->
-                jobs += async {
-                    GcAndroidTestMatrix.build(
-                        appApkGcsPath = apks.first,
-                        testApkGcsPath = apks.second,
-                        runGcsPath = runGcsPath,
-                        androidDeviceList = androidDeviceList,
-                        testShardsIndex = testShardsIndex,
-                        args = androidArgs,
-                        shardCounter = shardCounter,
-                        toolResultsHistory = history
-                    ).executeWithRetry()
+        apks.forEach { apk ->
+            // ensure we only shard tests that are part of the test apk
+            val testShardChunks = AndroidTestShard.getTestShardChunks(androidArgs, apk.test)
+            allTestShardChunks += testShardChunks
+            repeat(runCount) {
+                testShardChunks.forEach { testTargets ->
+                    jobs += async {
+                        GcAndroidTestMatrix.build(
+                            appApkGcsPath = apk.app ?: androidArgs.appApk,
+                            testApkGcsPath = apk.test,
+                            runGcsPath = runGcsPath,
+                            androidDeviceList = androidDeviceList,
+                            testTargets = testTargets,
+                            args = androidArgs,
+                            shardCounter = shardCounter,
+                            toolResultsHistory = history
+                        ).executeWithRetry()
+                    }
                 }
             }
         }
 
-        GenericTestRunner.afterRunTests(jobs.awaitAll(), runGcsPath, stopwatch, androidArgs)
+        println(GenericTestRunner.beforeRunMessage(androidArgs, allTestShardChunks))
+        val matrixMap = GenericTestRunner.afterRunTests(jobs.awaitAll(), runGcsPath, stopwatch, androidArgs)
+        matrixMap to allTestShardChunks
     }
 
     /**
      * Upload APKs if the path given is local
      *
      * @return Pair(gcs uri for app apk, gcs uri for test apk)
      */
-    private suspend fun resolveApks(args: AndroidArgs, runGcsPath: String): Pair<String, String> = coroutineScope {
+    private suspend fun resolveApks(args: AndroidArgs, runGcsPath: String): List<AppTestPair> = coroutineScope {
         val gcsBucket = args.resultsBucket
+        val appTestApks = listOf(AppTestPair(app = args.appApk, test = args.testApk)) + args.additionalAppTestApks
+        val result = mutableListOf<AppTestPair>()
 
-        val appApkGcsPath = async {
-            if (!args.appApk.startsWith(FtlConstants.GCS_PREFIX)) {
-                GcStorage.uploadAppApk(args, gcsBucket, runGcsPath)
-            } else {
-                args.appApk
-            }
-        }
+        appTestApks.forEach { apks ->
+            val appApkGcsPath = async { GcStorage.upload(apks.app ?: args.appApk, gcsBucket, runGcsPath) }
```

Comment:
Same here we should specify a `Dispatcher`

Reply:
Fixed